### PR TITLE
Include all translatable content in Stringsdict format

### PR DESCRIPTION
**Pull request by rigaspapas**
Make the Stringsdict format parser transform the source file in order to include all the translatable content in the produced strings.

Checklist (for the reviewer)
----------------------------

* [ ] Problem and solution are well-explained in the PR
* [ ] Change is covered by unit-tests
* [ ] Code is well documented
* [ ] Code is styled well and is following best practices
* [ ] Performs well when applied to big organizations/resources/etc from our production database
* [ ] Errors are handled properly
* [ ] All (conceivable) edge-cases are handled
* [ ] Code is instrumented to report unexpected failures or increases in the failure rate
* [ ] Commits have been squashed so that each one has a clear purpose (and green tests)
* [ ] Proper labels have been applied

Problem
-------
If my source looks like this:

```
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>TEST</key>
	<dict>
		<key>NSStringLocalizedFormatKey</key>
		<string>repeats %#@week_frequency@ on Monday</string>
		<key>week_frequency</key>
		<dict>
			<key>NSStringFormatSpecTypeKey</key>
			<string>NSStringPluralRuleType</string>
			<key>NSStringFormatValueTypeKey</key>
			<string>d</string>
			<key>one</key>
			<string>weekly</string>
			<key>other</key>
			<string>every %d weeks</string>
		</dict>
	</dict>
</dict>
</plist>
```

...then the following rules appear in the editor:
Form "one": weekly
Form "other": every %d weeks

This means that the "repeats [...] on Monday" part is never included and translated, but should be.

Steps to reproduce
------------------
Use a stringsdict file that contains translatable content in the top level of the plural rules, like this:

```
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>TEST</key>
	<dict>
		<key>NSStringLocalizedFormatKey</key>
		<string>repeats %#@week_frequency@ on Monday</string>
		<key>week_frequency</key>
		<dict>
			<key>NSStringFormatSpecTypeKey</key>
			<string>NSStringPluralRuleType</string>
			<key>NSStringFormatValueTypeKey</key>
			<string>d</string>
			<key>one</key>
			<string>weekly</string>
			<key>other</key>
			<string>every %d weeks</string>
		</dict>
	</dict>
</dict>
</plist>
```

Solution
--------

Transform the template, moving the translatable content *in* the plural rules, like this:
```      <key>NSStringLocalizedFormatKey</key>
		<string>%#@week_frequency@</string>
		<key>week_frequency</key>
		<dict>
			<key>NSStringFormatSpecTypeKey</key>
			<string>NSStringPluralRuleType</string>
			<key>NSStringFormatValueTypeKey</key>
			<string>d</string>
			<key>one</key>
			<string>repeats weekly on Monday</string>
			<key>other</key>
			<string>repeats every %d weeks on Monday</string>
		</dict>
```

Example run / Screenshots
-------------------------

Performance on live data
------------------------

### CODE REVIEW

File: openformats/formats/stringsdict.py [hunk 1]
```diff
@@ -90,34 +131,73 @@ def _handle_child_pairs(self, key_tag, dict_tag):
...
+        text_extras = None
+
         for key_child in dict_iterator:
             # The second key contains the secondary key.
             secondary_key = self._handle_key(key_child)
             value_tag = self._get_key_value(dict_iterator, key_child)
             if secondary_key == self.KEY_FORMAT:
                 # If the key is the one of the stringsdict defaults skip it
+                matches = re.match(self.KEY_FORMAT_RE, value_tag.content)
+                if matches is not None:
```
Comment (dbendilas) [inline openformats/formats/stringsdict.py]: The #if comment is not exactly correct any more at this point, as it refers to the `continue` statement a few lines below.

File: openformats/formats/stringsdict.py [hunk 2]
```diff
@@ -129,13 +209,25 @@ def _handle_strings(self, dict_tag, main_key, secondary_key):
...
             rule_number = self._validate_plural(key_content, key_tag)
-            strings_dict[rule_number] = value_tag.content
+            # If transformation [1] is needed, and we now process the FIRST
+            # variable, we apply the prefix and the suffix to each of the
+            # plural rules
+            if extras is not None and extras["variable"] == secondary_key:
+                content = "{prefix}{content}{suffix}".format(
+                    prefix=extras["prefix"],
+                    content=value_tag.content,
+                    suffix=extras["suffix"]
```
Comment (dbendilas) [inline openformats/formats/stringsdict.py]: Please add a trailing comma here.

File: openformats/formats/stringsdict.py [hunk 3]
```diff
@@ -27,11 +28,14 @@ class StringsDictHandler(Handler):
...
     PARSE_START = "<dict"
 
-    # Ignored keys
+    # Ignored or special keys
     KEY_FORMAT = 'NSStringLocalizedFormatKey'
     KEY_SPEC = 'NSStringFormatSpecTypeKey'
     KEY_VALUE = 'NSStringFormatValueTypeKey'
 
+    # Regular expressions
+    KEY_FORMAT_RE = r'(?P<prefix>[^@]*)%#@(?P<variable>\w+)@(?P<suffix>.*)'
```
Comment (dbendilas) [inline openformats/formats/stringsdict.py]: This regex is not associated with a key, but a value, so I think it needs better naming.
